u5-rest-api/features/covidController.js

```diff
@@ -0,0 +1,142 @@
+const { db } = require('../_services/firebase-admin-initialized')
+const { firestore } = require('firebase-admin')
+
+const readCovidRecords = async (req, res) => {
+  try {
+  // 1. input
+  // none
+
+    // 2. Query
+    const query = db.collection('covid-latest').get()
+
+    // 3. Response
+    const payload = (await query)
+      .docs
+      .map(doc => doc.data())
+      .map(({ date, stateId, stateName, cases, casesNew, vaccineOne, vaccineOnePercent, vaccineComplete, vaccineCompletePercent }) =>
+        ({ date: date.toMillis(), stateId, stateName, cases, casesNew, vaccineOne, vaccineOnePercent, vaccineComplete, vaccineCompletePercent }))
+
+    res.json({
+      result: 'ok',
+      payload,
+      count: payload.length
+    })
+  } catch (err) {
+    console.error(err)
+    res.status(500).json({
+      result: 'error',
+      payload: [],
+      count: 0
+    })
+  }
+}
+const readCovidRecord = async (req, res) => {
+  try {
+    // 1. input
+    const stateId = req.params.stateId.toUpperCase()
+
+    // query
+    const query = db.collection('covid-history').doc(stateId).get()
+
+    // Response
+    const snapshot = await query
+    if (!snapshot.exists) { return res.status(404).json({ response: 'not found' }) }
+
+    const data = snapshot.data()
+    const payload = {
+      stateId,
+      stateName: data.stateName,
+      history: data.history.map(({ date, cases, casesNew, vaccineOne, vaccineOnePercent, vaccineComplete, vaccineCompletePercent }) =>
+        ({
+          date: date.toMillis(),
+          cases,
+          casesNew,
+          vaccineOne,
+          vaccineOnePercent,
+          vaccineComplete,
+          vaccineCompletePercent
+        }))
+
+    }
+    res.json({
+      result: 'ok',
+      payload,
+      count: payload.length
+    })
+  } catch (err) {
+    console.error(err)
+    res.status(500).json({
+      result: 'error',
+      payload: [],
+      count: 0
+    })
+  }
+}
+const createCovidRecord = async (req, res) => {
+  try {
+    // 1. Inputs
+    const { stateId, date, cases, casesNew, vaccineOne, vaccineOnePercent, vaccineComplete, vaccineCompletePercent } = req.body
+    const record = {
+      stateId,
+      date: firestore.Timestamp.fromMillis(date),
+      cases,
+      casesNew,
+      vaccineOne,
+      vaccineOnePercent,
+      vaccineComplete,
+      vaccineCompletePercent
+    }
+    const historyRecord = {
+      history: firestore.FieldValue.arrayUnion({
+        date: firestore.Timestamp.fromMillis(date),
+        cases,
+        casesNew,
+        vaccineOne,
+        vaccineOnePercent,
+        vaccineComplete,
+        vaccineCompletePercent
+      })
+    }
```

Comment:
FYI: ฉันพบวิธีที่ชาญฉลาดในการจัดการการแปลงรูปแบบของ "date" ฉันชอบเพราะชื่อตัวแปรมีความชัดเจนมากกว่า และฉันรู้สึกสบายใจที่จะเขียนบรรทัดสุดท้ายในบรรทัดเดียว:

FYI: I found a clever way to manage converting the format of "date". I prefer it because the variable names are clearer, and I feel more comfortable writing the last line on a single line:

```js
const { stateId, date: dateMillis, cases, ... } = req.body // เปลี่ยนชื่อ date เป็น dateMillis
const date = firestore.Timestamp.fromMillis(dateMillis)
const record = { stateId, date, cases, ... } // ทั้งหมดเป็นแบบสั้น
```